packages/react/dropdown-menu/src/DropdownMenu.tsx

```diff
@@ -226,37 +226,34 @@ const DropdownMenuRootContent = React.forwardRef((props, forwardedRef) => {
       {...contentProps}
       ref={forwardedRef}
       portalled={portalled}
-      onCloseAutoFocus={composeEventHandlers(props.onCloseAutoFocus, (event) => {
-        event.preventDefault();
-        if (!preventCloseAutoFocusRef.current) context.triggerRef.current?.focus();
-      })}
-      onEscapeKeyDown={composeEventHandlers(props.onEscapeKeyDown, () => {
-        preventCloseAutoFocusRef.current = false;
-      })}
-      onPointerDownOutside={composeEventHandlers(
-        props.onPointerDownOutside,
-        (event) => {
-          const originalEvent = event.detail.originalEvent;
-          const ctrlLeftClick = originalEvent.button === 0 && originalEvent.ctrlKey === true;
-          const isRightClick = originalEvent.button === 2 || ctrlLeftClick;
-          preventCloseAutoFocusRef.current = isRightClick || !context.modal;
-        },
-        { checkForDefaultPrevented: false }
-      )}
-      onInteractOutside={composeEventHandlers(
-        props.onInteractOutside,
-        (event) => {
-          // Prevent dismissing when clicking the trigger.
-          // As the trigger is already setup to close, without doing so would
-          // cause it to close and immediately open.
-          //
-          // We use `onInteractOutside` as some browsers also
-          // focus on pointer down, creating the same issue.
-          const targetIsTrigger = context.triggerRef.current?.contains(event.target as HTMLElement);
-          if (targetIsTrigger) event.preventDefault();
-        },
-        { checkForDefaultPrevented: false }
-      )}
+      onCloseAutoFocus={(event) => {
+        props.onCloseAutoFocus?.(event);
+
+        if (!event.defaultPrevented && !isInteractOutsideRef.current) {
+          event.preventDefault();
+          if (!isInteractOutsideRef.current) context.triggerRef.current?.focus();
+        }
+
+        isInteractOutsideRef.current = false;
+      }}
+      onInteractOutside={(event) => {
+        props.onInteractOutside?.(event);
+
+        // Prevent dismissing when clicking the trigger.
+        // As the trigger is already setup to close, without doing so would
+        // cause it to close and immediately open.
+        //
+        // We use `onInteractOutside` as some browsers also
+        // focus on pointer down, creating the same issue.
+        const target = event.target as HTMLElement;
+        const targetIsTrigger = context.triggerRef.current?.contains(target);
+        if (targetIsTrigger) event.preventDefault();
+        else if (!event.defaultPrevented) {
+          const originalEvent = event.detail.originalEvent as PointerEvent;
+          const isModalLeftClick = context.modal && originalEvent.button === 0;
+          isInteractOutsideRef.current = !isModalLeftClick;
```

Comment:
I'm still struggling to follow the double negatives a bit, like we set a `isInteractOutside` which is about what "happened" and set it to the negative of something.

Then later we do `if (!isInteractOutside…)`

I feel like it would be easier for the ref we track to be about the "what" rather than the "how", ie. something like `shouldReturnFocus`.

Can we go on a call and pair quickly to see if it would help?